src/plugins/formulas/formulas.js

```diff
@@ -88,356 +38,297 @@ export class Formulas extends BasePlugin {
+    });
 
-    this.sheet.addLocalHook('afterRecalculate', (...args) => this.onSheetAfterRecalculate(...args));
+    this.sheetName = this.hyperformula.addSheet();
 
-    super.enablePlugin();
-  }
+    this.addHook('afterLoadData', (...args) => this.onAfterLoadData(...args));
+    this.addHook('modifyData', (...args) => this.onModifyData(...args));
+    this.addHook('modifySourceData', (...args) => this.onModifySourceData(...args));
 
-  /**
-   * Disables the plugin functionality for this Handsontable instance.
-   */
-  disablePlugin() {
-    super.disablePlugin();
-  }
+    // TODO test if the `before` hook will actually block operations
+    this.addHook('beforeCreateRow', (...args) => this.onBeforeCreateRow(...args));
+    this.addHook('beforeCreateCol', (...args) => this.onBeforeCreateCol(...args));
 
-  /**
-   * Returns cell value (evaluated from formula expression) at specified cell coords.
-   *
-   * @param {number} row Row index.
-   * @param {number} column Column index.
-   * @returns {*}
-   */
-  getCellValue(row, column) {
-    const cell = this.sheet.getCellAt(row, column);
+    this.addHook('afterCreateRow', (...args) => this.onAfterCreateRow(...args));
+    this.addHook('afterCreateCol', (...args) => this.onAfterCreateCol(...args));
 
-    return cell ? (cell.getError() || cell.getValue()) : void 0;
-  }
+    this.addHook('beforeRemoveRow', (...args) => this.onBeforeRemoveRow(...args));
+    this.addHook('beforeRemoveCol', (...args) => this.onBeforeRemoveCol(...args));
 
-  /**
-   * Checks if there are any formula evaluations made under specific cell coords.
-   *
-   * @param {number} row Row index.
-   * @param {number} column Column index.
-   * @returns {boolean}
-   */
-  hasComputedCellValue(row, column) {
-    return this.sheet.getCellAt(row, column) !== null;
-  }
+    this.addHook('afterRemoveRow', (...args) => this.onAfterRemoveRow(...args));
+    this.addHook('afterRemoveCol', (...args) => this.onAfterRemoveCol(...args));
 
-  /**
-   * Recalculates all formulas (an algorithm will choose the best method of calculation).
-   */
-  recalculate() {
-    this.sheet.recalculate();
-  }
+    // Autofill hooks
+    {
+      // Scoped into this block instead of being on the whole class to prevent
+      // other places from messing with it.
+      const lastAutofillSource = { value: undefined };
+
+      // Abuse the `modifyAutofillRange` hook to get the autofill start coordinates.
+      this.addHook('modifyAutofillRange', (_, entireArea) => {
+        const [startRow, startCol, endRow, endCol] = entireArea;
+
+        lastAutofillSource.value = {
+          start: {
+            row: startRow,
+            col: startCol
+          },
+          end: {
+            row: endRow,
+            col: endCol
+          }
+        };
+      });
+
+      // Abuse this hook to easily figure out the direction of the autofill
+      this.addHook('beforeAutofillInsidePopulate', (index, direction, _input, _deltas, _, selected) => {
```

Comment:
I get it that we wanted to use what we have. But it may not be the best idea.

The order of hooks is: modifyAutoFillRange, beforeAutofill, beforeAutofillInsidePopulate, afterAutofill.

And I see few issues here:
- You use `modifyAutoFillRange` to get the start coordinates and store it internally. As a user I register my `modifyAutoFillRange` hook and modify (as the name says) the start coordinates. Your stored value is outdated now.
- As a user I register `beforeAutofill` and return `false`, I'm blocking the autofill action. The start coordinates are not cleared, they're stored internally expecting the action to finish. Probably will be overwritten on the next autofill, but till then this is a wrong state that may be error prone in other places
- As a user I register `beforeAutofillInsidePopulate` and return `false` to block the populate action. Your `beforeAutofillInsidePopulate` hook already finished and applied the changes to the HF, while mine has blocked applying it to the HOT.
- As a user I register `beforeAutofillInsidePopulate` and return different `result.value`, your `beforeAutofillInsidePopulate` hook has already pasted the old values to the HF.

I need second opinion on this @jansiegel

Reply:
I could see the way I've done it being a problem for the reasons you describe, but what would an idiomatic alternative to this be?

Reply:
Gather your requirements and redesign `beforeAutofill` and `afterAutofill`? @jansiegel ?

Reply:
@wojciechczerniak @nikersify All valid concerns IMO. What are the problems with using `beforeAutofill` and `afterAutofill` as they're designed now, though?

BTW, is using HF's `copy` and `paste` in this context a good idea, in terms of unwanted side-effects? For example:

- user <kbd>Ctrl+C</kbd>s a cell containing `A`
- user performs autofill on cell containing `B`
   - HF performs `copy`
   - HF performs `paste`
- user <kbd>Ctrl+V</kbd>s

What will get pasted? I'm guessing `B`, right?

Reply:
Nothing should be pasted. Autofill (cell edit, column move, any user action) should clear the clipboard. Excel works like this. So we should HF `copy()` > HF `paste()` > HF `clearClipboard()` here.

Reply:
@wojciechczerniak you're right - that's surprising though. In Google Sheets autofill seems to have no effect on copy/paste logic.

Reply:
One potential solution for 1. 2. and 3. could be to (somehow) force all these registered hooks to run as last in their respective chains?

Reply:
@jansiegel The problem with `beforeAutofill` and `afterAutofill` is that it does not have the starting range, it's not always a single cell. I think it would be helpful not only for HF. Any other algorithm (from Excel) need the "source" to figure out the target sequence. The mentioned in code `direction` also may be useful for us and others.

BTW. I think that the argument order is wrong. First is `start` which is IMO useless. With `beforeAutofill` you would want to modify the `data`, not coordinates. IF you could edit the coordinate, but you can return only the start. Ofc this can be done by reference I think.

@nikersify That would be a new precedence. Developers expect their hooks to be called after ours. This way they can modify the behaviour. If we force our hooks last, their modifications would be always overwritten.

Reply:
Something like
```ts
beforeAutofill(data: CellValue[][], startRange: Range, finalRange: Range): CellValue[][] | boolean
```
and maybe direction could be figured out from `startRange` and `finalRange`? Depending where is the `finalRange.end.row` and `finalRange.end.column` are relative to `startRange`? Should be tested tho if figuring all directions is possible

Reply:
> That would be a new precedence. Developers expect their hooks to be called after ours. This way they can modify the behaviour. If we force our hooks last, their modifications would be always overwritten.

@wojciechczerniak Don't we want them to run last anyway? We wouldn't override anything, as we'd get the last value of `modifyAutofillRange`, after all user modifications anyway. Same with `beforeAutofillInsidePopulate`, forcing our hook to run last would let any user-hook to block it if necessary, wouldn't it?

Also I'm confused about 4), you say `result.value` when referring to `beforeAutofillInsidePopulate`, yet I see no `result` variable anywhere in the hook's docs?

Reply:
Ok, to make it super clear:

I think I chose `beforeAutofillInsidePopulate` because it was the only hook that did provide `direction` and individual cell inputs, since that calculation seems to already be happening inside the autofill plugin itself.

I would need a hook that would provide the `entireArea` from `modifyAutofillRange` (also, it seems like that one should be called `startArea` and vice versa?) - basically the area that the whole autofill should be sourced from, the target area - `selected` from `beforeAutofillInsidePopulate`, and maybe `direction` for convenience so that I wouldn't have to compute it again in this plugin.

Such hook should somehow respect all the hook rules so that it's still cancellable/mutable by end users of course.

Reply:
> Also I'm confused about 4), you say result.value when referring to beforeAutofillInsidePopulate, yet I see no result variable anywhere in the hook's docs?

Take a look here:
https://github.com/handsontable/handsontable/blob/fa685dc8c194c0269b893d202580ce7091ed8322/src/core.js#L894-L901

The hook may return `result` if it has a key `value` the value that is later used by `setDataAtCell` will be overwritten. So what you will `paste` may be outdated because it was modified during the lifecycle of Handsontable.

Reply:
@jansiegel How would the value overriding work if we’re not gonna use `beforeAutofillInsidePopulate`’s `result.value`?

Reply:
Oh, I thought we were keeping that one, but just to replace the values, not get the direction? We could make the `before-` hook return the data, but I dunno...

Reply:
@jansiegel I think the point is to sync those two hooks up into one so that if one gets canceled the other won’t have to work with stale values. Maybe just stuff `entireArea` into `beforeAutofillInsidePopulate`? What’s the problem with that? @wojciechczerniak

Reply:
Work continues at https://github.com/handsontable/handsontable/pull/7987